### PR TITLE
autonomous mavic

### PR DESCRIPTION
**Pull request by joachimhgg**
This PR aims to develop autonomous drones patroling over the forest, detecting and dropping water on wildfires.

An issue appeared with the method `ComputeMonoplotting()`. It aims to compute the world coordinates from the image coordinates of the fire. It gets the image coordinates of the fire and return its world coordinates. If its works in some situations, the robot goes sometimes away the fire instead of going in, or not close enough; due to wrong given world coordinates. The method is based on http://sar.kangwon.ac.kr/etc/rs_note/rsnote/cp9/cp9-6.htm taken from https://stackoverflow.com/questions/68195056/finding-the-real-world-coordinates-of-an-object-from-a-camera 

To overtake this issue, I used an alternative solution which can be found into the code.

### CODE REVIEW

File: projects/forest_firefighters/controllers/fire/fire.py [hunk 1]
```diff
@@ -229,40 +245,44 @@ def checkExtinction(self, tree):  # check and extinct the fire if there is water
                         tree.stopFire()
 
     def run(self):
+        flag_start_sim = False
```
Comment (omichel) [inline projects/forest_firefighters/controllers/fire/fire.py]: I would rename this variable as `start_fire_now`.

File: projects/forest_firefighters/protos/Smoke.proto [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+              1 1
+            ]
+          }
+          coordIndex [
+            2 0 1 3 -1
+          ]
+          texCoordIndex [
+            2 0 1 3 -1
+          ]
+       }
```
Comment (omichel) [inline projects/forest_firefighters/protos/Smoke.proto]:
```suggestion
        }
```

File: projects/forest_firefighters/worlds/.forest_firefighters.wbproj [hunk 3]
```diff
@@ -5,14 +5,8 @@ sceneTreePerspectives: 000000ff0000000100000002000000c0000000fc01000000060100000
...
-robotWindow: robot
-globalOptionalRendering: CoordinateSystem
+globalOptionalRendering: CoordinateSystem::RadarFrustums::CameraFrustums::ContactPoints
 consoles: Console:All:All
-renderingDevicePerspectives: Mavic 2 PRO:camera;1;0.6625;0;0
-renderingDevicePerspectives: Spot:left flank camera;0;0.790741;0;0
-renderingDevicePerspectives: Spot:left head camera;0;0.790741;0;0
-renderingDevicePerspectives: Spot:rear camera;0;0.790741;0;0
-renderingDevicePerspectives: Spot:right flank camera;0;0.790741;0;0
-renderingDevicePerspectives: Spot:right head camera;0;0.790741;0;0
```
Comment (omichel) [inline projects/forest_firefighters/worlds/.forest_firefighters.wbproj]: It doesn't seem good to commit these changes.

File: projects/forest_firefighters/worlds/forest_firefighters.wbt [hunk 4]
```diff
@@ -223,17 +223,87 @@ SquareManhole {
...
+SquareManhole {
+  translation 5.64 2.93 16.98
+  rotation 1 0 0 1.57079633
+  name "manhole(2)"
+}
 Mavic2Pro {
   translation 4.36 4.12 16.79
   rotation 0 0 1 0
-  controller "mavic"
+  controller "autonomous_mavic_v4"
```
Comment (omichel) [inline projects/forest_firefighters/worlds/forest_firefighters.wbt]: I would remove the `_v4` suffix in the final version.

File: projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+        Monoplotting based on https://stackoverflow.com/questions/68195056/finding-the-real-world-coordinates-of-an-object-from-a-camera
+        and http://sar.kangwon.ac.kr/etc/rs_note/rsnote/cp9/cp9-6.htm
+        Parameters: 
+        coord_fire (list): x,y image coordinates of the fire
+        verbose (bool): whether to print status messages or not
+        """
+        resolution = self.camera.getWidth(), self.camera.getHeight()
+        roll = self.curr_pos[3]
+        pitch = self.curr_pos[4]
+        yaw = self.curr_pos[5]
```
Comment (omichel) [inline projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py]: This is the roll, pitch, yaw of the IMU, the ones of the camera may be different.

File: projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+        
+        Z_obj = 30
+
+        #intrinsic properties of the camera
+        cx, cy, cz = self.curr_pos[0:3] #principal point in the image ~= real world coordinates of the camera
```
Comment (omichel) [inline projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py]: This is the position of the GPS, not necessarily the one of the camera.

File: projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+        f = self.camera.getFocalLength() #0, not working
+
+        fieldOfView = self.camera.getFov() #0.78398rad FOV = 2*arctan(x/2f)
+        diag_img = np.sqrt(resolution[0]**2 + resolution[1]**2)
+        f = diag_img/(2*np.tan(fieldOfView/2)) #f = focal length
```
Comment (omichel) [inline projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py]: That seems wrong, the focal length doesn't depend on the image resolution in pixels.

File: projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+        resolution = self.camera.getWidth(), self.camera.getHeight()
+        roll = self.curr_pos[3]
+        pitch = self.curr_pos[4]
+        yaw = self.curr_pos[5]
+        
+        Z_obj = 30
+
+        #intrinsic properties of the camera
+        cx, cy, cz = self.curr_pos[0:3] #principal point in the image ~= real world coordinates of the camera
+        f = self.camera.getFocalLength() #0, not working
```
Comment (omichel) [inline projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py]: Yes, this is documented. It returns 0 if the camera has no Focus node.

**Review by omichel — COMMENTED** [review]
I found several possible sources for the error, see my comments.

File: projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py (hunk 7 above)
Comment (joachimhgg) [reply projects/forest_firefighters/controllers/autonomous_mavic_v4/autonomous_mavic_v4.py]: That's true. The focal length depends of the diagonal of the film or sensor, but it should be in m and not in pixel. Is it possible to get the size of a pixel in m or the diagonal size of the sensor?

File: projects/forest_firefighters/README.md [hunk 9]
```diff
@@ -77,3 +77,11 @@ The keyboard can be used to move the robot and drop water on the forest fire.
 
 ## Spot Demo Controller
 The Spot controller is a basic controller that just moves the legs of the robot. The keyboard can also be used to throw water on the forest fire 
+
+## Autonomous mavic Controller
```
Comment (omichel) [inline projects/forest_firefighters/README.md]:
```suggestion
## Autonomous Mavic Controller

```

File: projects/forest_firefighters/README.md [hunk 10]
```diff
@@ -77,3 +77,11 @@ The keyboard can be used to move the robot and drop water on the forest fire.
 
 ## Spot Demo Controller
 The Spot controller is a basic controller that just moves the legs of the robot. The keyboard can also be used to throw water on the forest fire 
+
+## Autonomous mavic Controller
+
+The autonomous mavic moves between chosen coordinates above a forest in order to patrol and detect a fire. The altitude and world coordinates can be chosen through the controller args. Each time a fire is detected, the drone goes above the fire and drop a quantity of water to stop the fire. The project is using OpenCV to detect the smoke color range, and a naive approach to move above the fire (moves until the fire is in the center of the image). An alternative approach could be to use monoplotting to compute the world coordinates from the image coordinates of the fire. (for example based on [this method](http://sar.kangwon.ac.kr/etc/rs_note/rsnote/cp9/cp9-6.htm) ).
```
Comment (omichel) [inline projects/forest_firefighters/README.md]: We usually write one sentence per line.
```suggestion
The autonomous Mavic drone moves between chosen coordinates above a forest in order to patrol and detect a fire.
The altitude and world coordinates can be chosen through the controller args.
Each time a fire is detected, the drone goes above the fire and drop a quantity of water to stop the fire.
The project is using OpenCV to detect the smoke color range, and a naive approach to move above the fire (moves until the fire is in the center of the image).
An alternative approach could be to use monoplotting to compute the world coordinates from the image coordinates of the fire (for example based on [this method](http://sar.kangwon.ac.kr/etc/rs_note/rsnote/cp9/cp9-6.htm)).  

```

File: projects/forest_firefighters/README.md [hunk 11]
```diff
@@ -77,3 +77,11 @@ The keyboard can be used to move the robot and drop water on the forest fire.
 
 ## Spot Demo Controller
 The Spot controller is a basic controller that just moves the legs of the robot. The keyboard can also be used to throw water on the forest fire 
+
+## Autonomous mavic Controller
+
+The autonomous mavic moves between chosen coordinates above a forest in order to patrol and detect a fire. The altitude and world coordinates can be chosen through the controller args. Each time a fire is detected, the drone goes above the fire and drop a quantity of water to stop the fire. The project is using OpenCV to detect the smoke color range, and a naive approach to move above the fire (moves until the fire is in the center of the image). An alternative approach could be to use monoplotting to compute the world coordinates from the image coordinates of the fire. (for example based on [this method](http://sar.kangwon.ac.kr/etc/rs_note/rsnote/cp9/cp9-6.htm) ).  
+
+It is also possible to add multiple drones to increase chances of detecting a fire. Please note that the fire is starting only when the drone have reached an altitude of 40m.
```
Comment (omichel) [inline projects/forest_firefighters/README.md]:
```suggestion
It is also possible to add multiple drones to increase chances of detecting a fire.
Please note that the fire is starting only when the drone has reached an altitude of 40m.
```

File: projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+    # Constants, empirically found.
+    K_VERTICAL_THRUST = 68.5  # with this thrust, the drone lifts.
+    K_VERTICAL_OFFSET = 0.6   # Vertical offset where the robot actually targets to stabilize itself.
+    K_VERTICAL_P = 3.0        # P constant of the vertical PID.
+    K_ROLL_P = 50.0           # P constant of the roll PID.
+    K_PITCH_P = 30.0          # P constant of the pitch PID.
+
+    MAX_YAW_DISTURBANCE = 0.4
+    MAX_PITCH_DISTURBANCE = -1
+    target_precison = 0.5 #precision between the target position and the robot position in meters
```
Comment (omichel) [inline projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py]: Please check PEP8 compliance. See https://github.com/cyberbotics/webots/wiki/Python-Coding-Style
```suggestion
    target_precision = 0.5  # precision between the target position and the robot position in meters
```

File: projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+
+    def __init__(self):
+        Robot.__init__(self)
+
+        self.timeStep = int(self.getBasicTimeStep())
+        
+        self.water_to_drop = 0
```
Comment (omichel) [inline projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py]: This is not consistent:
```suggestion
        self.time_step = int(self.getBasicTimeStep())
        
        self.water_to_drop = 0
```

File: projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+        self.rear_right_motor = self.getDevice("rear right propeller")
+        self.camera_pitch_motor = self.getDevice("camera pitch")
+        self.camera_pitch_motor.setPosition(1.55) #vertical PoV
+        motors = [self.front_left_motor, self.front_right_motor, self.rear_left_motor, self.rear_right_motor]
+        for motor in motors:
+            motor.setPosition(float('inf'))
+            motor.setVelocity(1)
+
+        self.curr_pos = 6*[0] # X,Y,Z, yaw, pitch, roll
+        self.target_pos = [0,0,0]
```
Comment (omichel) [inline projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py]: Please do not use abbreviations, they are often confusing as here:
```suggestion
        self.current_pose = 6*[0] # X,Y,Z, yaw, pitch, roll
        self.target_position = [0,0,0]
```

File: projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+        #turn the robot to the left or to the right according the value and the sign of angle_left
+        yaw_disturbance = self.MAX_YAW_DISTURBANCE*angle_left/(2*np.pi)
+        pitch_disturbance = clamp(np.log10(abs(angle_left)), self.MAX_PITCH_DISTURBANCE, 0.1) #non proportional and decruising function
+
+        if verbose_movement: 
+            distance_left = np.sqrt(((self.target_pos[0] - self.curr_pos[0]) **2) + ((self.target_pos[1] - self.curr_pos[1]) **2))
+            print("remaning angle: {:.4f}, remaning distance: {:.4f}".format(angle_left, distance_left))
+        return yaw_disturbance, pitch_disturbance
+
+    def NaiveApproach(self, verbose=True):
```
Comment (omichel) [inline projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py]: Consistency:
```suggestion
    def naive_approach(self, verbose=True):
```

File: projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+        """
+        Naive approch to move the robot above the fire. Closed loop to move the robot towards to the fire by step until to be above it.
```
Comment (omichel) [inline projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py]:
```suggestion
        Naive approach to move the robot above the fire. Closed loop to move the robot towards to the fire step-by-step until it reaches the fire.
```

File: projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+        pitch_disturbance = self.world_fire_pos[1]*clamp(abs(y_img-resolutionY/2),0,abs(self.MAX_PITCH_DISTURBANCE))
+
+        if self.world_fire_pos == [0,0]:             
+            self.water_to_drop = 15
+            if verbose: print("Water dropped on fire target: {} at position {}".format(self.target_pos[0:2], self.curr_pos[0:2]))
+            self.img_coord_fire = []
+
+        return yaw_disturbance, pitch_disturbance
+
+    def FireDetection(self, verbose=True):
```
Comment (omichel) [inline projects/forest_firefighters/controllers/autonomous_mavic/autonomous_mavic.py]:
```suggestion
    def fire_detection(self, verbose=True):
```

File: projects/forest_firefighters/worlds/forest_firefighters.wbt [hunk 18]
```diff
@@ -223,19 +223,41 @@ SquareManhole {
...
+  name "manhole(1)"
+}
 Mavic2Pro {
   translation 4.36 4.12 16.79
   rotation 0 0 1 0
-  controller "mavic"
+  controller "autonomous_mavic"
+  controllerArgs [
+    "--target_altitude=45"
+    "--patrol_coord=11 11, 21 11, 21 21, 11 21"
```
Comment (omichel) [inline projects/forest_firefighters/worlds/forest_firefighters.wbt]:
```suggestion
    "--patrol_coords=11 11, 21 11, 21 21, 11 21"
```

**Review by omichel — CHANGES_REQUESTED** [review]
I believe we should not remove the spot robot from the simulation.
Also, there are only two drones, can't we have 4?

**Comment by omichel** [comment]
Thank you, the code looks better. What about my original comments:

> I believe we should not remove the spot robot from the simulation.
> Also, there are only two drones, can't we have 4?

**Comment by omichel** [comment]
Thank you. I believe we should improve this simulation with the following:
- [x] Don't display the camera overlays.
- [x] Fix the depth fighting problem with the Mavic platforms (move them slightly higher).
- [x] Fix the 'D' key control for the Spot robot.
- [x] Restart random fires repeatedly so that we can see the Mavic robots in action more than once.
- [x] Make a small movie showing the Mavic robots fighting fires.

**Comment by joachimhgg** [comment]
> Thank you. I believe we should improve this simulation with the following:
> 
>     * [x]  Don't display the camera overlays.
> 
>     * [x]  Fix the depth fighting problem with the Mavic platforms (move them slightly higher).
> 
>     * [x]  Fix the 'D' key control for the Spot robot.
> 
>     * [x]  Restart random fires repeatedly so that we can see the Mavic robots in action more than once.
> 
>     * [x]  Make a small movie showing the Mavic robots fighting fires.

I fixed the different issues and added a video to the Readme

**Review by omichel — DISMISSED** [review]
All good. Thank you!

**Review by omichel — CHANGES_REQUESTED** [review]
It seems the robot window for controlling the wind is broken on this branch...

**Review by omichel — APPROVED** [review]
Looks good to me now.
Thank you.

**Pull request merged**